src/main/java/com/google/devtools/build/lib/bazel/bzlmod/Discovery.java

```diff
@@ -49,7 +52,9 @@ private Discovery() {}
    */
   @Nullable
   public static ImmutableMap<ModuleKey, InterimModule> run(
-      Environment env, RootModuleFileValue root)
+      Environment env,
+      RootModuleFileValue root,
+      ImmutableList.Builder<ImmutableMap<String, Optional<Checksum>>> nestedRegistryFileHashes)
```

Comment:
can we make `run` return a record instead?

also, IMO we should just make it one single map (i.e. merge eagerly). I don't see the eager merge being a noticeable perf hit, but the readability improvement would be quite noticeable IMO. (Also no need to check for mismatched checksums really ... `putIfAbsent` is good enough.)

Reply:
I made the result a record and am merging all maps eagerly now. I kept the explicit `Map` merge function though as the case where we do end up with different checksums for the same file would just be too obscure with `putIfAbsent`: module resolution could silently yield a different result during the next cold build since the different module files fetched during the first build would settle on the first in encounter order in the cold build. I would rather have a clean server crash in that case that users are likely to take to GitHub.

Reply:
do you foresee any other setup that causes this, other than "two --registry flags that only differ by a terminating slash and with contents that change during a single build"? If that's the only possible cause for differing checksums, I'm not sure it's worth _any_ complexity budget because it's such a contrived setup, and we could fix it on the registry parsing side by adding/stripping terminating slashes anyway.

Reply:
Agreed, I replaced the logic with `putAll` and am instead uniquifying registries now.